framework/src/application/node/synchronizer/base_synchronizer.ts

```diff
@@ -12,43 +12,123 @@
+	'EVENT_SYNCHRONIZER_SYNC_RQUIRED';
 
 export abstract class BaseSynchronizer {
-	protected logger: Logger;
-	protected channel: InMemoryChannel;
+	public events: EventEmitter;
+
+	protected _logger: Logger;
+	protected _channel: InMemoryChannel;
+	protected _chain: Chain;
 
-	public constructor(logger: Logger, channel: InMemoryChannel) {
-		this.logger = logger;
-		this.channel = channel;
+	public constructor(logger: Logger, channel: InMemoryChannel, chain: Chain) {
+		this._logger = logger;
+		this._channel = channel;
+		this._chain = chain;
+		this.events = new EventEmitter();
 	}
 
 	protected async _applyPenaltyAndRestartSync(
 		peerId: string,
-		receivedBlock: BlockInstance,
+		receivedBlock: Block,
 		reason: string,
 	): Promise<void> {
-		this.logger.info(
+		this._logger.info(
 			{ peerId, reason },
 			'Applying penalty to peer and restarting synchronizer',
 		);
-		await this.channel.invoke('app:applyPenaltyOnPeer', {
+		await this._channel.invoke('app:applyPenaltyOnPeer', {
 			peerId,
 			penalty: 100,
 		});
-		this.channel.publish('app:chain:sync', {
+		this.events.emit(EVENT_SYNCHRONIZER_SYNC_RQUIRED, {
 			block: receivedBlock,
+			peerId,
+		});
+	}
+
+	protected async _getLastBlockFromNetwork(
+		peerId: string,
+	): Promise<Block<BlockHeaderAsset>> {
+		const { data } = await this._channel.invokeFromNetwork<{
+			data: string | undefined;
+		}>('requestFromPeer', {
+			procedure: 'getLastBlock',
+			peerId,
+		});
+		if (!data || !data.length) {
+			throw new ApplyPenaltyAndRestartError(
+				peerId,
+				"Peer didn't provide its last block",
+			);
+		}
+		return this._chain.dataAccess.decode<BlockHeaderAsset>(
+			Buffer.from(data, 'base64'),
+		);
+	}
+
+	protected async _getHighestCommonBlockFromNetwork(
+		peerId: string,
+		ids: Buffer[],
+	): Promise<BlockHeader<BlockHeaderAsset>> {
+		const { data } = await this._channel.invokeFromNetwork<{
+			data: string | undefined;
+		}>('requestFromPeer', {
+			procedure: 'getHighestCommonBlock',
+			peerId,
+			data: {
+				ids: ids.map(id => id.toString('base64')),
+			},
 		});
+		if (!data || !data.length) {
+			throw new ApplyPenaltyAndAbortError(
+				peerId,
+				"Peer didn't return a common block",
+			);
+		}
+		return this._chain.dataAccess.decodeBlockHeader<BlockHeaderAsset>(
+			Buffer.from(data, 'base64'),
+		);
+	}
+
+	protected async _getBlocksFromNetwork(
+		peerId: string,
+		fromID: Buffer,
+	): Promise<Block<BlockHeaderAsset>[]> {
+		const { data } = await this._channel.invokeFromNetwork<{
+			data: string[] | undefined;
+		}>('requestFromPeer', {
+			procedure: 'getBlocksFromId',
+			peerId,
+			data: {
+				blockId: fromID.toString('base64'),
+			},
+		}); // Note that the block matching lastFetchedID is not returned but only higher blocks.
+		if (!data || !data.length) {
+			throw new Error('Peer did not respond with block');
+		}
+		return data.map(d =>
+			this._chain.dataAccess.decode<BlockHeaderAsset>(Buffer.from(d, 'base64')),
```

Comment:
When we are specifying return type here `<BlockHeaderAsset>`, then how can it don't require schema for that type?

Reply:
This is more like a type helper. the asset mapping happen by the version number registered to the chain library.
There was no better way to enable type and registering